### PR TITLE
API refresh - part two (jextract edition)

### PR DESCRIPTION
**Pull request by mcimadamore**
This patch fixes jextract to work with the latest API changes.

The patch reintroduces overloads for foreign function wrappers which take a scope (since ResourceScope is no longer a subtype of SegmentAllocator).

We will consider later whether to drop these overloads; it is likely that the majority of jextract clients will already have an allocator somewhere, so the extra code is not necessary. On the other hand, if we want overloads, then I think it would also be useful to add an overload that takes a `MemorySegment` and converts it into a `prefixAllocator` (so that clients can store result of foreign function which returns a struct-by-value in a pre-existing segments).

I've made some improvements to the string conversion logic; in a lot of places we were creating a scope, then doing a malloc to allocate a temporary CXString, then extract a Java String from the CXString, and dispose/free the CXString. Since jextract is single-threaded, I replaced all that with an high-order function which takes a segment supplier, and uses a prefix allocator under the hood. This should be way more efficient than what we had - and string conversion is quite frequent in libclang (e.g. `CXCursor::spelling`).

The patch passes all tests.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Reviewers
 * [Jorn Vernee](https://openjdk.java.net/census#jvernee) (@JornVernee - Committer) ⚠️ Review applies to fff0fc6cc27b24f9ec666daadc540df43c8f8d46

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/panama-foreign pull/594/head:pull/594` \
`$ git checkout pull/594`

Update a local copy of the PR: \
`$ git checkout pull/594` \
`$ git pull https://git.openjdk.java.net/panama-foreign pull/594/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 594`

View PR using the GUI difftool: \
`$ git pr show -t 594`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/panama-foreign/pull/594.diff">https://git.openjdk.java.net/panama-foreign/pull/594.diff</a>

</details>

### CODE REVIEW

File: src/jdk.incubator.jextract/share/classes/jdk/internal/clang/LibClang.java [hunk 1]
```diff
@@ -71,16 +77,24 @@ public static Index createIndex(boolean local) {
...
         return str;
     }
 
+    /**
+     * This is an allocator for temporary CXString structs. CXStrToString needs to save the CXString somewhere,
+     * so that we can extract a Java string out of it. Once that's done, we can dispose the CXString, and the
+     * associated segment. Since jextract is single-threaded, we can use a prefix allocator, to speed up string
+     * conversion. The size of the prefix segment is set to 256, which should be enough to hold a CXString.
+     */
+    private final static SegmentAllocator stringAllocator = SegmentAllocator.prefixAllocator(
```
Comment (PaulSandoz) [inline src/jdk.incubator.jextract/share/classes/jdk/internal/clang/LibClang.java:94]:
```suggestion
    private final static SegmentAllocator STRING_ALLOCATOR = SegmentAllocator.prefixAllocator(
```

File: src/jdk.incubator.jextract/share/classes/jdk/internal/clang/LibClang.java [hunk 2]
```diff
@@ -44,8 +47,11 @@
     private static final boolean CRASH_RECOVERY = Boolean.getBoolean("libclang.crash_recovery");
     private static final boolean IS_WINDOWS = System.getProperty("os.name").startsWith("Windows");
 
+    final static SegmentAllocator IMPLICIT_ALLOCATOR =
```
[the remark this answers is not in the document]
Comment (mcimadamore) [reply src/jdk.incubator.jextract/share/classes/jdk/internal/clang/LibClang.java:50]: nice suggestion

File: src/jdk.incubator.jextract/share/classes/jdk/internal/clang/LibClang.java [hunk 3]
```diff
@@ -71,16 +77,24 @@ public static Index createIndex(boolean local) {
         return index;
     }
 
-    public static String CXStrToString(MemorySegment cxstr) {
+    public static String CXStrToString(Function<SegmentAllocator, MemorySegment> segmentSupplier) {
+        MemorySegment cxstr = segmentSupplier.apply(stringAllocator);
```
Comment (JornVernee) [inline src/jdk.incubator.jextract/share/classes/jdk/internal/clang/LibClang.java:81]: I don't see `stringAllocator` anywhere, I guess this should be `STRING_ALLOCATOR`.
```suggestion
        MemorySegment cxstr = segmentSupplier.apply(STRING_ALLOCATOR);
```

**Pull request closed without merging**